interp/src/primitives.rs

```diff
@@ -869,6 +893,23 @@ impl StdMemD3 {
             update: None,
         }
     }
+
+    pub fn initialize_memory(&mut self, vals: &[Value]) {
+        assert_eq!(
+            (self.d0_size * self.d1_size * self.d2_size) as usize,
+            vals.len()
+        );
+
+        for (idx, val) in vals.iter().enumerate() {
+            assert_eq!(val.len(), self.width as usize);
+            self.data[idx] = val.clone()
+        }
+    }
+
+    #[inline]
+    fn calc_addr(&self, addr0: u64, addr1: u64, addr2: u64) -> u64 {
+        addr2 + self.d2_size * (addr1 + addr0 * self.d1_size)
```

Comment:
Also LGTM! This is harder to reverse, of course, but if you want:

`addr0 * self.d1_size * self.d2_size + addr1 * self.d2_size + addr2`

…sorta has a rhythm to it.